packages/apollo-gateway/src/index.ts

```diff
@@ -126,11 +147,56 @@ export class ApolloGateway implements GraphQLService {
     this.isReady = true;
   }
 
+  public onSchemaChange(value: SchemaChangeCallback): Unsubscriber {
+    if (!isManagedConfig(this.config)) {
+      return () => {};
+    }
+
+    this.onSchemaChangeListeners.add(value);
+    if (!this.pollingTimer) this.startPollingServices();
+
+    return () => {
+      this.onSchemaChangeListeners.delete(value);
+      if (this.onSchemaChangeListeners.size === 0 && this.pollingTimer) {
+        clearInterval(this.pollingTimer!);
+        this.pollingTimer = undefined;
+      }
+    };
+  }
+
+  private startPollingServices() {
+    if (this.pollingTimer) clearInterval(this.pollingTimer);
+
+    this.pollingTimer = setInterval(async () => {
+      try {
+        const [services, isNewSchema] = await this.loadServiceDefinitions(
+          this.config,
+        );
+        if (!isNewSchema) {
+          this.logger.debug('No changes to gateway config');
+          return;
+        }
+        if (this.queryPlanStore) this.queryPlanStore.flush();
+        this.logger.debug('Gateway config has changed, updating schema');
+        this.createSchema(services);
+        this.onSchemaChangeListeners.forEach(listener =>
+          listener(this.schema!),
```

Comment:
I wonder what the use cases for having `onSchemaChangeListeners` be `async` might be?  Might one of these need to hold up the actual roll-over?  Should the `createSchema` be changed after all have successfully been invoked?  (Or is it okay to change our own internal representation of the `schema` first, as it is right now.)

Reply:
Are you proposing this becomes
```js
await Promise.all(this.onSchemaChangeListeners.map(listener => listener(this.schema!)))
```
?

Reply:
We can't pass them a new schema until we call `createSchema` to create it. Unless we make `createSchema` return a schema and we assign it later?